[stability] sql: add special tracing to trap #7881

andreimatei opened the pull request:
sql: create root spans for transactions

Also:

I've decided to go medieval and invest in tracing all SQL execution in
the hope that #7881 will be triggered again.
The tracing is gated by an env var. When turned on, every SQL txn
creates a new tracer that accumulates spans in the session's txnState
(this is similar to how "SQL tracing" currently works). When we detect
7881, we mark the root span of the current one of these traces for
"sampling", which means that later, when that span (and hence the trace)
is closed, we dump the trace with all the log messages in it (note that
currently this trace only has one span, since we're not very good at
starting child spans yet).

cc @knz, @cockroachdb/stability

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8681)

<!-- Reviewable:end -->

andreimatei committed:
sql: add special tracing to trap #7881

I've decided to go medieval and invest in tracing all SQL execution in
the hope that #7881 will be triggered again. This builds on the prev
commit that added the ability to collect spans across the executor and a
client.Txn.
The tracing is gated by an env var. When turned on, every SQL txn
creates a new tracer that accumulates spans in the session's txnState
(this is similar to how "SQL tracing" currently works). When we detect
7881, we mark the root span of the current one of these traces for
"sampling", which means that later, when that span (and hence the trace)
is closed, we dump the trace with all the log messages in it (note that
currently this trace only has one span, since we're not very good at
starting child spans yet).

andreimatei committed:
sql: move span collection from a client.Txn to a TxnState

The TxnState matches the lifetime of a SQL txn, whereas the client.Txn
goes away when the SQL txn goes, for example, to the Aborted state.
Plus, the SQL executor has direct access to a txnState, so it's natural
for it to use the TxnState's context for log messages, and we want these
messages aassocited with the "transaction's trace". In fact, the
TxnState was already the the holder of a txn's root span, so this is in
effect unifying the owner of the span and of all the collected children
spans.

This commit also makes all SQL transactions be executed with a root
span.

andreimatei committed:
log: add file:line to trace log messages

This only affects messages coming from log.Infof(). Messages going
directly to log.Trace(), or messages going through log.VTrace() but with
the verbosity turned off are not affected and will continue to not have
file/line.

petermattis commented:
Reviewed 4 of 4 files at r1, 1 of 1 files at r2.
Review status: 1 of 5 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[internal/client/txn.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3HF70nKn1H53m545:-KPY3HF8uDyTbXt8SO2q:b-3ocy5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/internal/client/txn.go#L103)):*
> ```Go
> 	// initiated on this host.
> 	// It's also used by "EXPLAIN TRACE".
> 	// Note that in SQL land there's also TxnState.CollectedSpans. That guys
> ```

`s/. That guys/which/g`

---

*[sql/executor.go, line 520 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4cteiFWB85glJAuB:-KPY4cteiFWB85glJAuC:boh2nn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L520)):*
> ```Go
> 			results, remainingStmts, err = runTxnAttempt(e, planMaker, origState, txnState, opt, stmtsToExec)
> 
> 			// Until #7881 fixed.
> ```

Add `TODO(andrei)`.

---

*[sql/executor.go, line 544 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY5XcPR3HSKPCQH9ri:-KPY5XcPR3HSKPCQH9rj:b-1tiirn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L544)):*
> ```Go
> 					err, txnState.txn.Proto, execOpt.AutoRetry, execOpt.AutoCommit, stmts,
> 					remainingStmts)
> 				// Until #7881 fixed.
> ```

`TODO(andrei)`

---

*[sql/executor.go, line 549 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY7E9RPbqAgQfpqF5u:-KPY7E9RPbqAgQfpqF5v:bc6iqkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L549)):*
> ```Go
> 						"txnState %+v, execOpt %+v, stmts %+v, remaining %+v",
> 						err, txnState, execOpt, stmts, remainingStmts)
> 					txnState.sp.SetBaggageItem(keyFor7881Sample, "sample me please")
> ```

Perhaps hide in a method: `txnState.forceTrace()`.

---

*[sql/session.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3dme4498ScWIxC5-:-KPY3dme4498ScWIxC50:b-5z23cf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L162)):*
> ```Go
> 	State TxnStateEnum
> 
> 	// baseCtx is used to init the contexts for every SQL txn, as this txnState is
> ```

`s/baseCtx/Ctx/g`

---

*[sql/session.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3whvoQz_Q6IyhMzV:-KPY3whvoQz_Q6IyhMzW:b-1hglpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L253)):*
> ```Go
> 
> 	if ts.txn != nil {
> 		ts.CollectedSpans = append(ts.CollectedSpans, ts.txn.CollectedSpans...)
> ```

Might be worth optimizing the common case where the transaction is only executed once:

```
if len(ts.CollectedSpans) == 0 {
  ts.CollectedSpans = tx.txn.CollectedSpans
} else {
  ts.CollectedSpans = append(ts.CollectedSpans, ts.txn.CollectedSpans...)
}
```

---

*[sql/session.go, line 55 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY71KexOxTawfA6IGg:-KPY71KfArsQZd4hc4bY:b-dub6ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/session.go#L55)):*
> ```Go
> 
> // span baggage key used for marking a span
> var keyFor7881Sample = "found#7881"
> ```

This seems overly specific to 7881. Perhaps `keyForceSample`.

---

*[sql/session.go, line 245 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY7RJ3dZ4xZ_cL54_b:-KPY7RJ3dZ4xZ_cL54_c:b-2bb3o4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/session.go#L245)):*
> ```Go
> 			opentracing.ContextWithSpan(ctx, sp), tr)
> 		// The new span will be Finish()ed by FinishSpan().
> 		ts.sp = sp
> ```

Are you not using a snowball trace because of the greater expense?

---

*[util/log/structured.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4KW5LYR9lleCc_05:-KPY4KW5LYR9lleCc_06:bksbjy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9cbf9668a4301536c894a62fe4ea6c79380d17c/util/log/structured.go#L63)):*
> ```Go
> 	file, line, _ := caller.Lookup(depth + 1)
> 	msg := makeMessage(ctx, format, args)
> 	Trace(ctx, file+":"+strconv.Itoa(line)+" "+msg)
> ```

Not sure if we want this. It isn't too hard to figure out which line generated a trace statement. Grep is your friend. @RaduBerinde thoughts?

---

*[util/tracing/tracer.go, line 94 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY5mKveHUCQTrQq1uq:-KPY5mKveHUCQTrQq1ur:b-7eko17) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/util/tracing/tracer.go#L94)):*
> ```Go
> 	sp, err := JoinOrNew(tr, nil, "sql txn")
> 	return sp, tr, err
> }
> ```

Seems like this belongs in `sql`, not here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Do we want to retain some of this infrastructure beyond 7881? If so we should change to more generic names

I don't understand one thing - is the intention that we would collect traces from other nodes involved in the txn and dump those too? Because that would only happen if we enable snowball tracing for that txn no? (or am I missing something?)

---

Review status: 1 of 5 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[sql/session.go, line 128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPYafrYUnhFDrP99GrZ:-KPYafrYUnhFDrP99Gr_:bo88q11) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L128)):*
> ```Go
> // reading descriptors consistent with the txn in which they'll be used?
> func (s *Session) Ctx() context.Context {
> 	if s.TxnState.sp != nil {
> ```

Seems odd that we check `sp` here. If `sp != nil` happens to mean that the transaction is active, move it into a `txnState.IsActive()` method

---

*[sql/session.go, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPYb3kSj2N4JDAWPQhV:-KPYb3kSj2N4JDAWPQhW:b-t0t9oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L164)):*
> ```Go
> 	// baseCtx is used to init the contexts for every SQL txn, as this txnState is
> 	// reset for every txn.
> 	Ctx context.Context
> ```

The comment is deceiving, "is used to init the contexts for every SQL txn" means it is some kind of generic context that is a parent for txn contexts, but from the code it seems it just is the transaction context.. I would rename to `txnCtx` and clarify the comment.

---

*[util/log/structured.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4KW5LYR9lleCc_05:-KPYcI_BFIDLRnkqiEUT:b-a2phtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9cbf9668a4301536c894a62fe4ea6c79380d17c/util/log/structured.go#L63)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Not sure if we want this. It isn't too hard to figure out which line generated a trace statement. Grep is your friend. @RaduBerinde thoughts?
</details>
The logs already have these and sometimes they're useful (especially if we are logging an error, we may find the place where the error is generated based on the message, but it might be hard to figure out which layer logged it). That said, I haven't worked too much with traces.. An option would also be to put them at the end of the message where they might be less disruptive

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KPYdesLUUe6b-wENN3p:b-nnuowt)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
(Thanks guys! Please take another look.)

Well there's no "infrastructure" here to write home about. You know I'm a fan of tracing in principle, and we should have a way to sample everything and be able to get a hold of the current trace. But I don't think this is it, by creating a new tracer per request in a random place and accumulating the trace state in a random object. We should all sit down with Tobi too and have a heart to heart.

I didn't enable snowballing because, according to some comment by Peter, that's the expensive part of "sql tracing", so I didn't want to enable if *for everything* (and there's no way to enable it per trace). Plus frankly I don't think it'd be very useful for finding this bug.

@raduberinde I've made a change in `AdminServe` about the contexts of sql sessions opened from RPCs. I don't really know what the deal with contexts that come from gRPC is, maybe you do. Please take a look.

---

Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[internal/client/txn.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3HF70nKn1H53m545:-KPYs6qXTKy5hjjIn176:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/internal/client/txn.go#L103)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/. That guys/which/g`
</details>
Done.

---

*[sql/executor.go, line 520 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4cteiFWB85glJAuB:-KPZEUFqYkWz99FcZBFX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L520)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Add `TODO(andrei)`.
</details>
Done.

---

*[sql/executor.go, line 544 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY5XcPR3HSKPCQH9ri:-KPYshiWcqbl5MLJ-oh3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L544)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `TODO(andrei)`
</details>
Done.

---

*[sql/executor.go, line 549 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY7E9RPbqAgQfpqF5u:-KPYszQGxdkwSU5yVDPG:b-jlsunz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L549)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Perhaps hide in a method: `txnState.forceTrace()`.
</details>
i dunno... I wouldn't pollute `txnState` more than I did already.

---

*[sql/session.go, line 128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPYafrYUnhFDrP99GrZ:-KPZ4-K0_Q7Z4uttdtFy:b-j10ld2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L128)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Seems odd that we check `sp` here. If `sp != nil` happens to mean that the transaction is active, move it into a `txnState.IsActive()` method
</details>
true. We're checking the right thing now. And I've renamed the functions that deal with the span to better suggest that the span is tied to a SQL txn's lifetime.

---

*[sql/session.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3dme4498ScWIxC5-:-KPZEXq3f7WZbifZfhOn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L162)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/baseCtx/Ctx/g`
</details>
Done.

---

*[sql/session.go, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPYb3kSj2N4JDAWPQhV:-KPZDyxRni0tVj8isLlx:b-q4r6bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L164)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The comment is deceiving, "is used to init the contexts for every SQL txn" means it is some kind of generic context that is a parent for txn contexts, but from the code it seems it just is the transaction context.. I would rename to `txnCtx` and clarify the comment.
</details>
I went the other way and actually made it a "base context". I made all transactions have a root span, and hence a modified child context. See if you dig.
For this, I've had to ensure that the a session's ctx always has a tracer in it. It seems reasonable to me...

---

*[sql/session.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3whvoQz_Q6IyhMzV:-KPZERIApyAdwyXDzCZw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L253)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Might be worth optimizing the common case where the transaction is only executed once:
> 
> ```
> if len(ts.CollectedSpans) == 0 {
>   ts.CollectedSpans = tx.txn.CollectedSpans
> } else {
>   ts.CollectedSpans = append(ts.CollectedSpans, ts.txn.CollectedSpans...)
> }
> ```
</details>
Done.

---

*[sql/session.go, line 55 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY71KexOxTawfA6IGg:-KPYt9JGMGZE2QqKPirH:b5hmjy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/session.go#L55)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This seems overly specific to 7881. Perhaps `keyForceSample`.
</details>
Radu also said something about using more generic names, but I feel like nothing done here is generic, and it all needs to go away, in favor of a more general method of "forced sampling" that works with the global singleton tracer we should be using.

---

*[sql/session.go, line 245 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY7RJ3dZ4xZ_cL54_b:-KPYo7c2kHkUFEsTC9JF:b-uz635w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/session.go#L245)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Are you not using a snowball trace because of the greater expense?
</details>
correct. I want to enable this on the beta cluster until I catch the sucker.

---

*[util/log/structured.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4KW5LYR9lleCc_05:-KPYpY5edScSGEMwBK2z:bexv34w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9cbf9668a4301536c894a62fe4ea6c79380d17c/util/log/structured.go#L63)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The logs already have these and sometimes they're useful (especially if we are logging an error, we may find the place where the error is generated based on the message, but it might be hard to figure out which layer logged it). That said, I haven't worked too much with traces.. An option would also be to put them at the end of the message where they might be less disruptive
</details>
To be honest, I think the file and line are the most interesting part of these log messages :) There were definitely some messages that just said "error: xxx" and I had no idea where that's coming from. What's the downside of having them? I didn't put them in `log.Trace()` because maybe there's a cost to walking the stack, but here?


Here's how a trace looks without them (and it's gonna continue to look the same for most messages since they're `log.Tracef()`s):
```
I160818 22:48:29.702194 sql/session.go:291  SQL trace:
     0.046ms      0.000ms    executing 1/1: BEGIN TRANSACTION
     0.161ms      0.115ms    executing 1/1: DROP TABLE t.kv
     0.206ms      0.045ms    meta descriptor lookup
     0.355ms      0.149ms    looked up range descriptor
     0.591ms      0.236ms    looked up range descriptor
     0.613ms      0.022ms    sending RPC to [/Min, /Table/11)
     0.747ms      0.134ms    meta descriptor lookup
     0.751ms      0.004ms    looked up range descriptor from cache
     0.759ms      0.008ms    sending RPC to [/Min, /Table/11)
     0.863ms      0.104ms    meta descriptor lookup
     0.867ms      0.003ms    looked up range descriptor from cache
     0.875ms      0.008ms    sending RPC to [/Min, /Table/11)
```

And here's one with it:
```
     3.788ms      2.420ms    sql/event_log.go:109 Event: "drop_table", target: 51, info: {TableName:kv Statement:DROP TABLE t.kv User:root}
```

---

*[util/tracing/tracer.go, line 94 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY5mKveHUCQTrQq1uq:-KPYqrgtSKGf66PG8bWL:b-xx64rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/util/tracing/tracer.go#L94)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Seems like this belongs in `sql`, not here.
</details>
I had it in sql initially, but it uses the `defaultOptions` which isn't public... So I'd just keep it in here. I do intend for it to go away... Added the TODO to remove.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KPYmHQEhvBOhqKTCmtu:b4pqxim)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Sure there is, Snowballing is enabled in a span (and its future children) by setting the snowball baggage item. In short, if you set that baggage item, requests that go out will return populated `CollectedSpans` (see https://github.com/cockroachdb/cockroach/blob/a6350417c9f6f368c8657b2edfc0e384d1c71310/server/node.go#L793)

---

Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KPZOjxCP2Xrx6iSFD6q:bshwjeq)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
oh I see, ok. But anyway, by the time I figure out a trace is interesting, it's too late for future snowballing to be useful.

---

Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KPZQyCp06ZO0ELgJEQU:bu042u1)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ah, gotcha

---

Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KPZRpwpmUUcve8j_7aX:b7j4s9c)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[server/admin.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZSdaLNNnIWi4oVyd8:-KPZSdaLNNnIWi4oVyd9:by2fxb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f25cc75b6bd257bc28e04e89cc55f7774b332c/server/admin.go#L172)):*
> ```Go
> 	ctx context.Context, args sql.SessionArgs,
> ) *sql.Session {
> 	ctx = tracing.WithTracer(ctx, tracing.TracerFromCtx(s.server.ctx.Ctx))
> ```

Nice, yeah this is in principle what we need to do - we need to "merge" information from an incoming context with information that is relevant to where the request got in now. We will eventually do the same with log tags and whatever else we store in context, but there is no infrastructure for that yet. Maybe add a `TODO(radu): figure out a general way to do this`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r4, 1 of 1 files at r5, 1 of 4 files at r6.
Review status: 4 of 7 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[server/admin.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZSdaLNNnIWi4oVyd8:-KPZrIM64YkV0oh3SPC7:b-em6deg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f25cc75b6bd257bc28e04e89cc55f7774b332c/server/admin.go#L172)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Nice, yeah this is in principle what we need to do - we need to "merge" information from an incoming context with information that is relevant to where the request got in now. We will eventually do the same with log tags and whatever else we store in context, but there is no infrastructure for that yet. Maybe add a `TODO(radu): figure out a general way to do this`
</details>
What's the point of creating a new context here? Just that all the information from the server's context is removed?

---

*[sql/executor.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZrVFPYdm2X5DhJVC2:-KPZrVFPYdm2X5DhJVC3:b-1zf0g9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd02fcc85634ab89e132319cd9a75e1eef346bb/sql/executor.go#L287)):*
> ```Go
> }
> 
> // NewDummyExecutor creates an empty Executor that is used for certain tests.
> ```

while you're here, can you put this in a _test.go file?

---

*[sql/executor.go, line 572 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZromL6Nd-J0Ij5yZ4:-KPZromL6Nd-J0Ij5yZ5:bsvcr05) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd02fcc85634ab89e132319cd9a75e1eef346bb/sql/executor.go#L572)):*
> ```Go
> 		}
> 
> 		// If we're no longer in a transaction, finish the trace.
> ```

does this want to be in a defer right after the txnState.ResetForSomethingSomething?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 4 of 7 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[server/admin.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZSdaLNNnIWi4oVyd8:-KPZvM_EE6Z_q6HBcmMx:b1pk9v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f25cc75b6bd257bc28e04e89cc55f7774b332c/server/admin.go#L172)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What's the point of creating a new context here? Just that all the information from the server's context is removed?
</details>
We want to inherit the gRPC context, we are just adding a value to it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 4 of 7 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[server/admin.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZSdaLNNnIWi4oVyd8:-KPZvmegA67YHPFqPmF8:bpdvf7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f25cc75b6bd257bc28e04e89cc55f7774b332c/server/admin.go#L172)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> We want to inherit the gRPC context, we are just adding a value to it
</details>
Oh, duh.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 7 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[server/admin.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZSdaLNNnIWi4oVyd8:-KPZxqanukCqp1nD4kXr:bhnbg3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f25cc75b6bd257bc28e04e89cc55f7774b332c/server/admin.go#L172)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Oh, duh.
</details>
added a TODO

---

*[sql/executor.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZrVFPYdm2X5DhJVC2:-KP_1UlvCNJsgvaBWzHD:bgfgcl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd02fcc85634ab89e132319cd9a75e1eef346bb/sql/executor.go#L287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> while you're here, can you put this in a _test.go file?
</details>
unfortunately it uses the private `cfg` field... I don't know if I should make that public...

---

*[sql/executor.go, line 572 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZromL6Nd-J0Ij5yZ4:-KPZzNWVp2miWtlhM0yO:b-3p3v17) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd02fcc85634ab89e132319cd9a75e1eef346bb/sql/executor.go#L572)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does this want to be in a defer right after the txnState.ResetForSomethingSomething?
</details>
It wants, but unfortunately it's not that easy. The problem is that SQL transactions can span ping-pongs back to the client, so different invocations of this method. So the state machine that we need to maintain needs to be more explicit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 6 files at r7, 1 of 1 files at r8, 3 of 4 files at r9.
Review status: 6 of 7 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZrVFPYdm2X5DhJVC2:-KPjrCA5rMfsg8bOx3Lo:bi0dxhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd02fcc85634ab89e132319cd9a75e1eef346bb/sql/executor.go#L287)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> unfortunately it uses the private `cfg` field... I don't know if I should make that public...
</details>
I meant that you can move this function to `sql/helpers_test.go` so that it is only accessible from tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 6 of 7 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZrVFPYdm2X5DhJVC2:-KPn-zx3iQMmM-9brx0a:bkju9rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd02fcc85634ab89e132319cd9a75e1eef346bb/sql/executor.go#L287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I meant that you can move this function to `sql/helpers_test.go` so that it is only accessible from tests.
</details>
the test using it is in package `pgwire`, so I think this function can't be in a `_test` within `sql`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 6 of 7 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 287 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPZrVFPYdm2X5DhJVC2:-KPn1hbe1tpOlDype2dS:bxecj2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd02fcc85634ab89e132319cd9a75e1eef346bb/sql/executor.go#L287)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> the test using it is in package `pgwire`, so I think this function can't be in a `_test` within `sql`.
</details>
Ah, that's true.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 6 of 7 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[util/log/structured.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4KW5LYR9lleCc_05:-KQRqI0u6RmZsBwPD33K:b-dhch8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9cbf9668a4301536c894a62fe4ea6c79380d17c/util/log/structured.go#L63)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> To be honest, I think the file and line are the most interesting part of these log messages :) There were definitely some messages that just said "error: xxx" and I had no idea where that's coming from. What's the downside of having them? I didn't put them in `log.Trace()` because maybe there's a cost to walking the stack, but here?
> 
> 
> Here's how a trace looks without them (and it's gonna continue to look the same for most messages since they're `log.Tracef()`s):
> ```
> I160818 22:48:29.702194 sql/session.go:291  SQL trace:
>      0.046ms      0.000ms    executing 1/1: BEGIN TRANSACTION
>      0.161ms      0.115ms    executing 1/1: DROP TABLE t.kv
>      0.206ms      0.045ms    meta descriptor lookup
>      0.355ms      0.149ms    looked up range descriptor
>      0.591ms      0.236ms    looked up range descriptor
>      0.613ms      0.022ms    sending RPC to [/Min, /Table/11)
>      0.747ms      0.134ms    meta descriptor lookup
>      0.751ms      0.004ms    looked up range descriptor from cache
>      0.759ms      0.008ms    sending RPC to [/Min, /Table/11)
>      0.863ms      0.104ms    meta descriptor lookup
>      0.867ms      0.003ms    looked up range descriptor from cache
>      0.875ms      0.008ms    sending RPC to [/Min, /Table/11)
> ```
> 
> And here's one with it:
> ```
>      3.788ms      2.420ms    sql/event_log.go:109 Event: "drop_table", target: 51, info: {TableName:kv Statement:DROP TABLE t.kv User:root}
> ```
> 
</details>
What's annoying is that the file/line may or may not show up depending how it makes it way into the event. E.g. if you use a `VEventf`, it depends on the verbosity whether the file/line shows up. But better to sometimes have more information than be consistent and never have it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KQRpkqgxTZ0GB-zXDNt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on sql/executor.go:
```diff
@@ -523,6 +523,15 @@ func (e *Executor) execRequest(session *Session, sql string) StatementResults {
 			var err error
 			results, remainingStmts, err = runTxnAttempt(e, planMaker, origState, txnState, opt, stmtsToExec)
+
+			// Until #7881 fixed.
```
TODO(andrei) here and elsewhere

tschottdorf commented:
Reviewed 1 of 7 files at r4, 6 of 6 files at r7, 7 of 7 files at r10, 1 of 1 files at r11, 5 of 5 files at r13, 4 of 4 files at r14.
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---

*[internal/client/txn.go, line 449 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFP3g9maIUsIYwG8XI:-KRFP3g9maIUsIYwG8XJ:b-wukhgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/internal/client/txn.go#L449)):*
> ```Go
> // considered finalized and cannot be used to send any more commands.
> func (txn *Txn) Rollback() error {
> 	log.VTracef(2, txn.Context, "rolling back Txn")
> ```

s/Txn/transaction/

---

*[internal/client/txn.go, line 560 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFP7FNbUV-j4Ar45BL:-KRFP7FNbUV-j4Ar45BM:bb7462z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/internal/client/txn.go#L560)):*
> ```Go
> 		err = fn(txn, &opt)
> 
> 		// Until 7881 is fixed.
> ```

TODO

---

*[internal/client/txn.go, line 569 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFP9LJPVSABktnt4Qk:-KRFP9LJPVSABktnt4Ql:bb7462z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/internal/client/txn.go#L569)):*
> ```Go
> 			// fn succeeded, but didn't commit.
> 			err = txn.Commit()
> 			log.Tracef(txn.Context, "client.Txn did AutoCommit. err: %s\ntxn: %+v", err, txn.Proto)
> ```

TODO

---

*[sql/executor.go, line 1069 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFLtBqFcalqobNnj56:-KRFLtBrr6c5MlWNVjrX:b23s9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/executor.go#L1069)):*
> ```Go
> 		case commit:
> 			// We're done with this txn.
> 			txnState.resetStateAndTxn(NoTxn)
> ```

This would always happen before `finishSQLTxn()`?

---

*[sql/executor.go, line 527 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFPLyU-jNFT5nkDRPB:-KRFPLyU-jNFT5nkDRPC:bb7462z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/sql/executor.go#L527)):*
> ```Go
> 			results, remainingStmts, err = runTxnAttempt(e, planMaker, origState, txnState, opt, stmtsToExec)
> 
> 			// Until #7881 fixed.
> ```

TODO

---

*[sql/executor.go, line 551 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFPSWBiL41GYpMQt7S:-KRFPSWBiL41GYpMQt7T:bkeenlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/sql/executor.go#L551)):*
> ```Go
> 					err, txnState.txn.Proto, execOpt.AutoRetry, execOpt.AutoCommit, stmts,
> 					remainingStmts)
> 				// TODO(andrei): Until #7881 fixed.
> ```

Is this log message above part of 7881? Because it looks pretty messy (I guess that's ok for now though if it helps).

---

*[sql/session.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3dme4498ScWIxC5-:-KRFMRzZgmPun_DMFy18:bh8hct1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L162)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done.
</details>
I don't understand this comment. As far as I can tell, `Ctx` is the base context to be used for anything that goes on in this transaction. Why doesn't it contain `sp` below (which then wouldn't have to be stored separately)?

---

*[sql/session.go, line 55 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY71KexOxTawfA6IGg:-KRFM-7-RZM_NKXVgfme:b-1s4jt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/session.go#L55)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Radu also said something about using more generic names, but I feel like nothing done here is generic, and it all needs to go away, in favor of a more general method of "forced sampling" that works with the global singleton tracer we should be using.
</details>
While I'm also not fond of this special-cased code, it's hidden behind an env var and can hopefully go away soon, and the explicit names make sure in the meantime that nobody confuses this for anything it isn't.

---

*[sql/session.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMAaYnkBS-tLmkIdS:-KRFMAaZe7kAhtJjIbd-:b-64y1it) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L67)):*
> context is the Session's base context, used by Ctx().

---

*[sql/session.go, line 193 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMdAB6SNL2PhBNr7F:-KRFMdAB6SNL2PhBNr7G:b-gkqed1) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L193)):*
> ```Go
> 
> 	sp opentracing.Span
> 	// When "SQL tracing" is enabled, CollectedSpans accumulates spans as they're
> ```

Perhaps you can give a more direct reference to what `SQL tracing` is.

---

*[sql/session.go, line 234 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFNS8S6tlo1b1RXgJ8:-KRFNS8S6tlo1b1RXgJ9:bntsv6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L234)):*
> ```Go
> 		// Create a root span for this SQL txn.
> 		tracer := tracing.TracerFromCtx(ctx)
> 		ts.sp = tracer.StartSpan(fmt.Sprintf("sql txn for user: %q", s.User))
> ```

just "sql transaction" sounds more idiomatic. This is supposed to be a "component" of the system; filtering by user would be done through tags (but I don't think the API/frontend for that is quite there yet).

---

*[util/log/structured.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4KW5LYR9lleCc_05:-KRFOiO9QbD7E8uq-ULC:b-s7e0dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9cbf9668a4301536c894a62fe4ea6c79380d17c/util/log/structured.go#L63)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> What's annoying is that the file/line may or may not show up depending how it makes it way into the event. E.g. if you use a `VEventf`, it depends on the verbosity whether the file/line shows up. But better to sometimes have more information than be consistent and never have it.
</details>
There might be a small performance cost associated with it, but I doubt we're going to care about that order of magnitude for a while and saving a few greps a day should be worth it for the time being.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
This isn't clear from the PR description, but what this PR accomplishes is actually quite useful: It'll allow us to trace complete SQL transactions in lightstep (where now the parts of the transaction which go through TxnCoordSender aren't part of a common root span, and thus aren't related to each other). This should be useful for getting a better idea of where transactions are spending their time, and how much they retry, and where.

andreimatei committed:
sql: clean up a transaction when the session closes

At the very least, we have to close the transaction's root span.
We also do more cleanup and attempt to rollback the txn. This is nice
because it lets the rollback attempt be done in the context of the txn.
It supersedes the aborting that was being done by the TxnCoordSender
through the cancellation of the session's context with interrupted the
txn's heartbeat loop.

andreimatei commented:
edited the PR description

---

Review status: all files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

*[internal/client/txn.go, line 449 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFP3g9maIUsIYwG8XI:-KRFxo8bcWBkjDwrC8ph:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/internal/client/txn.go#L449)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Txn/transaction/
</details>
Done.

---

*[internal/client/txn.go, line 560 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFP7FNbUV-j4Ar45BL:-KRFxx6xJz_xPMbTvBSF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/internal/client/txn.go#L560)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> TODO
</details>
Done.

---

*[internal/client/txn.go, line 569 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFP9LJPVSABktnt4Qk:-KRFy5e1J2Dz7CniXEsB:byxl7nl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/internal/client/txn.go#L569)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> TODO
</details>
I think this can stay

---

*[sql/executor.go, line 1069 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFLtBqFcalqobNnj56:-KRFlfKiZWkJrBTX6Egn:b-8w254r) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/executor.go#L1069)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This would always happen before `finishSQLTxn()`?
</details>
well, this or a rollback. `finishSQLTxn()` is called when the something moves the TxnState to the NO_TXN state.

We have a problem with not-so-well-behaved clients that don't end their transactions properly. Similar to what I've found yesterday, about closing of pg connections that didn't properly roll back.
I've moved that fix to this PR, in a new commit. Now `session.Finish` properly cleans up in that case, including closing the span.

---

*[sql/executor.go, line 527 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPVmtmKB9wPd0DZDvgk-r9-527:-KRFyHQ3O5nzy0vHS4WE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23b042bed51af4964933b5306e882e140e79e7ce/sql/executor.go#L527)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> TODO(andrei) here and elsewhere 
</details>
Done.

---

*[sql/executor.go, line 551 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFPSWBiL41GYpMQt7S:-KRFlP3txxtaiFgJ9y3v:b2nih0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/sql/executor.go#L551)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Is this log message above part of 7881? Because it looks pretty messy (I guess that's ok for now though if it helps).
</details>
yup

---

*[sql/session.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY3dme4498ScWIxC5-:-KRFjJxXHXFkU1CFYHw7:bvphyq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381ca3f8e4f26661046d2e23046669c0f25dcf7/sql/session.go#L162)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't understand this comment. As far as I can tell, `Ctx` is the base context to be used for anything that goes on in this transaction. Why doesn't it contain `sp` below (which then wouldn't have to be stored separately)?
</details>
reworded comment. And removed `sp`.

---

*[sql/session.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMAaYnkBS-tLmkIdS:-KRFd4BC0rn_RwdlrrxR:bd7ozor) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L67)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> > context is the Session's base context, used by Ctx().
</details>
meh, I like my form better

---

*[sql/session.go, line 193 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMdAB6SNL2PhBNr7F:-KRFj_StvCkxTiklEnG_:bb70u45) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L193)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Perhaps you can give a more direct reference to what `SQL tracing` is.
</details>
replaced with the env var name

---

*[sql/session.go, line 234 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFNS8S6tlo1b1RXgJ8:-KRFjvdEw2ojQMgaJIg8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L234)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> just "sql transaction" sounds more idiomatic. This is supposed to be a "component" of the system; filtering by user would be done through tags (but I don't think the API/frontend for that is quite there yet).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KRFyvH0vFAVxPHoIu7I:b-rnrych)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 6 of 6 files at r15, 1 of 1 files at r16.
Review status: 2 of 8 files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

*[sql/session.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMAaYnkBS-tLmkIdS:-KRGBGdOSnRriogTfwCC:b-ugc58n) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L67)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> meh, I like my form better
</details>
Ambiguous phrases like "You probably don't want to"  aren't really fit for usage in code. At least say "Don't use it directly; see Ctx()`.

---

*[sql/session.go, line 130 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRGBd0vwD6NJgnzRP_K:-KRGBd0vwD6NJgnzRP_L:bkv7wio) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5000350ad5246287a504c236c9a67021db2c64d/sql/session.go#L130)):*
> ```Go
> 	// If we're inside a txn, roll it back.
> 	if s.TxnState.State != NoTxn && s.TxnState.State != Aborted {
> 		s.TxnState.updateStateAndCleanupOnErr(
> ```

won't this put us in `NoTxn` state so that we don't call `finishSQLTxn` below even though we should?

---

*[sql/session.go, line 138 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRGBlgtypSqXGYx2Qyb:-KRGBlgtypSqXGYx2Qyc:b3yjngd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5000350ad5246287a504c236c9a67021db2c64d/sql/session.go#L138)):*
> ```Go
> 	// This will stop the heartbeating of the of the txn record.
> 	// TODO(andrei): This shouldn't have any effect, since, if there was a
> 	// transaction, we just explicitly rolled it back above, so the hearbeat loop
> ```

s/hearbeat/heartbeat/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 8 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 520 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4cteiFWB85glJAuB:-KRKBfrPK6xBcyS8bxxm:bpjmb0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L520)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done.
</details>
I don't see anything changed here.

---

*[util/tracing/tracer.go, line 94 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY5mKveHUCQTrQq1uq:-KRKBVHHRtjrHrIqEAYi:b-exj68k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/util/tracing/tracer.go#L94)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I had it in sql initially, but it uses the `defaultOptions` which isn't public... So I'd just keep it in here. I do intend for it to go away... Added the TODO to remove.
</details>
You could export `defaultOptions`. Guess I don't feel strongly either way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KRKC6iSCIHTMJPlDNhP:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> only nits.

---

Reviewed 2 of 2 files at r17, 4 of 4 files at r18.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[internal/client/txn.go, line 569 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRKPeVdOmCjUuxLYfpR:-KRKPeVdOmCjUuxLYfpS:bocvm38) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c7774da42acf25533c73d3424cff443ceb2036b/internal/client/txn.go#L569)):*
> ```Go
> 			// fn succeeded, but didn't commit.
> 			err = txn.Commit()
> 			log.Tracef(txn.Context, "client.Txn did AutoCommit. err: %s\ntxn: %+v", err, txn.Proto)
> ```

use %v for the error as it can be nil.

---

*[sql/executor.go, line 551 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFPSWBiL41GYpMQt7S:-KRKPmD8IjiUGDqIQ_Z8:b-fkayxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/sql/executor.go#L551)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> yup
</details>
Then the TODO should be above. You can put both statements in a block. Also, the error should be logged with %s (and `aErr`, not `err`).

---

*[sql/session.go, line 56 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRKQ6-AgjE1vGke9RLJ:-KRKQ6-AgjE1vGke9RLK:b-pye2iz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c7774da42acf25533c73d3424cff443ceb2036b/sql/session.go#L56)):*
> ```Go
> 
> // span baggage key used for marking a span
> var keyFor7881Sample = "found#7881"
> ```

s/var/const/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KRKQJIOZBr36wiryzt9:bomf0a7)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
TFTR guys!

---

Review status: 2 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[internal/client/txn.go, line 569 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRKPeVdOmCjUuxLYfpR:-KRKZQd6LNzJa-bpOPkx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c7774da42acf25533c73d3424cff443ceb2036b/internal/client/txn.go#L569)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> use %v for the error as it can be nil.
</details>
Done.

---

*[sql/executor.go, line 551 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFPSWBiL41GYpMQt7S:-KRK_akI_mEGD3Dtlpmv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b18570109eb667019ed6a409c54dc696163d8790/sql/executor.go#L551)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Then the TODO should be above. You can put both statements in a block. Also, the error should be logged with %s (and `aErr`, not `err`).
</details>
Done.

---

*[sql/session.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMAaYnkBS-tLmkIdS:-KRK_i5gN-UXyEqtzXL6:b-saxkfr) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L67)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Ambiguous phrases like "You probably don't want to"  aren't really fit for usage in code. At least say "Don't use it directly; see Ctx()`.
</details>
well except sometimes you do want to use `s.context`, if you know what you're doing. The phrase means pay attention.

---

*[sql/session.go, line 130 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRGBd0vwD6NJgnzRP_K:-KRKaBst8FupPYm60Q51:b-fihd6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5000350ad5246287a504c236c9a67021db2c64d/sql/session.go#L130)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> won't this put us in `NoTxn` state so that we don't call `finishSQLTxn` below even though we should?
</details>
no, it's gonna put us in `Aborted` state.

---

*[sql/session.go, line 138 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRGBlgtypSqXGYx2Qyb:-KRKbNUBGIOz8SvnG1Au:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5000350ad5246287a504c236c9a67021db2c64d/sql/session.go#L138)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/hearbeat/heartbeat/
</details>
Done.

---

*[sql/session.go, line 56 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRKQ6-AgjE1vGke9RLJ:-KRK_feHIydAHzbSt9SY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c7774da42acf25533c73d3424cff443ceb2036b/sql/session.go#L56)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/var/const/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-:-KRKdrI0rNlPGW8dDIH8:b-q2baky)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 8 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[sql/executor.go, line 520 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KPY4cteiFWB85glJAuB:-KRKfYInsGHZ0qUoWbHy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1bd1b3e3e8545bf4e1537c08769639f12dda534/sql/executor.go#L520)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I don't see anything changed here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 6 of 6 files at r19, 2 of 2 files at r21, 4 of 4 files at r22.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[internal/client/txn.go, line 569 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRKPeVdOmCjUuxLYfpR:-KRKiMeogbfSflKno2_V:b-7v3a4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c7774da42acf25533c73d3424cff443ceb2036b/internal/client/txn.go#L569)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done.
</details>
`%v`.

---

*[sql/session.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMAaYnkBS-tLmkIdS:-KRKgZXDSFUTV1cMFXqW:b-abh2ut) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L67)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well except sometimes you do want to use `s.context`, if you know what you're doing. The phrase means pay attention.
</details>
I'm not even going to entertain that argument. May I remind you that you're defending a comment that reads "You probably don't want to use it directly".
BTW Tamir thinks this whole comment is useless (including the reference to `Ctx()`) and that the relevant comment is already on `Ctx()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[internal/client/txn.go, line 569 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRKPeVdOmCjUuxLYfpR:-KRKvmjxBLrUQLqpb_Aj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c7774da42acf25533c73d3424cff443ceb2036b/internal/client/txn.go#L569)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `%v`.
</details>
Done.

---

*[sql/session.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681#-KRFMAaYnkBS-tLmkIdS:-KRKvnaAZCeunjjDg3dU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de71544a57a2b8ea1741c71cabd02dc8e6f69478/sql/session.go#L67)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I'm not even going to entertain that argument. May I remind you that you're defending a comment that reads "You probably don't want to use it directly".
> BTW Tamir thinks this whole comment is useless (including the reference to `Ctx()`) and that the relevant comment is already on `Ctx()`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 6 of 6 files at r23, 2 of 2 files at r25, 4 of 4 files at r26.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8681)*
<!-- Sent from Reviewable.io -->
